src/analysis/seminorm.lean

```diff
@@ -0,0 +1,219 @@
+lemma absorbent_nhds_zero (hA : A ∈ nhds (0 : E)) : absorbent 𝕜 A :=
+begin
+  intro x,
+  rcases mem_nhds_sets_iff.mp hA with ⟨w, hw₁, hw₂, hw₃⟩,
+  have hc : continuous (λ t : 𝕜, t • x), from continuous_id.smul continuous_const,
+  rcases metric.is_open_iff.mp (hc _ hw₂) 0 (by rwa [mem_preimage, zero_smul])
+    with ⟨r, hr₁, hr₂⟩,
+  have hr₃, from inv_pos.mpr (half_pos hr₁),
+  use [(r/2)⁻¹, hr₃],
+  intros a ha₁,
+  have ha₂ : 0 < ∥a∥, from calc 0 < _ : hr₃ ... ≤ _ : ha₁,
+  have ha₃ : a ⁻¹ • x ∈ w, begin
+    apply hr₂,
+    rw [metric.mem_ball, dist_eq_norm, sub_zero, norm_inv],
+    calc ∥a∥⁻¹ ≤ r/2 : (inv_le (half_pos hr₁) ha₂).mp ha₁
+    ...       < r : half_lt_self hr₁,
+  end,
+  rw [mem_smul_set_iff_inv_smul_mem (norm_pos_iff.mp ha₂)],
+  exact hw₁ ha₃,
+end
+
+/-- The union of `{0}` with the interior of a balanced set
+    is balanced. -/
+lemma balanced_zero_union_interior (hA : balanced 𝕜 A) :
+  balanced 𝕜 ({(0 : E)} ∪ interior A) :=
+begin
+  intros a ha, by_cases a = 0,
+  { rw [h, zero_smul_set],
+    apply subset_union_left _,
+    exact union_nonempty.mpr (or.intro_left _ $ singleton_nonempty _) },
+  { rw [←image_smul, image_union, union_subset_iff], split,
+    { rw [image_singleton, smul_zero], exact subset_union_left _ _ },
+    { apply subset_union_of_subset_right,
+      apply interior_maximal,
+      rw image_subset_iff,
+      calc _ ⊆ A : interior_subset
+      ...    ⊆ _ : by { rw ←image_subset_iff, exact hA _ ha},
+      exact is_open_map_smul_of_ne_zero h _ is_open_interior }},
+end
+
+/-- The interior of a balanced set is balanced if it contains the origin. -/
+lemma balanced_interior (hA : balanced 𝕜 A) (h : (0 : E) ∈ interior A) :
+  balanced 𝕜 (interior A) :=
+begin
+  rw ←singleton_subset_iff at h,
+  rw [←union_eq_self_of_subset_left h],
+  exact balanced_zero_union_interior hA,
+end
+
+/-- The closure of a balanced set is balanced. -/
+lemma balanced_closure (hA : balanced 𝕜 A) : balanced 𝕜 (closure A) :=
+begin
+  intros a ha,
+  calc _ ⊆ closure (a • A) :
+  by { simp_rw ←image_smul,
+       exact image_closure_subset_closure_image
+               (continuous_const.smul continuous_id) }
+  ...    ⊆ _ : closure_mono (hA _ ha),
+end
+
+end
+
+
+/-- A seminorm on a vector space over a normed field is a function to
+the reals that is positive semidefinite, positive homogeneous, and
+subadditive. -/
+
+structure seminorm (𝕜 : Type*) (E : Type*)
+  [normed_field 𝕜] [add_comm_group E] [vector_space 𝕜 E] :=
+(to_fun   : E → ℝ)
+(smul     : ∀ (a : 𝕜) (x : E), to_fun (a • x) = ∥a∥ * to_fun x)
+(triangle : ∀ x y : E, to_fun (x + y) ≤ to_fun x + to_fun y)
+
+variables
+{𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [add_comm_group E] [vector_space 𝕜 E]
+
+instance : inhabited (seminorm 𝕜 E) :=
+⟨{ to_fun   := λ _, 0,
+   smul     := λ _ _, (mul_zero _).symm,
+   triangle := λ x y, by rw add_zero }⟩
+
+instance : has_coe_to_fun (seminorm 𝕜 E) := ⟨_, λ p, p.to_fun⟩
+
+section seminorm
```

Comment:
Since `seminorm` is a `structure` (not a `class`) and it will be an explicit argument in all these lemmas, it makes sense to move them to the namespace `seminorm`. Then you can drop `seminorm_` in (some) names.